### DISCUSSION TITLE
Can't we just substitute application code for built-in repl code?

### DISCUSSION THREAD

**Discussion by guest271314** [nodejs/single-executable#15] [Ideas]
When `node` is run at the command line we get a REPL session

> If we run the node command without any script to execute or without any arguments, we start a REPL session:
> 
>    node

Since it is unlikely an application will be using the built-in REPL can't we just substitute application code for the built-in REPL code, so that all an end-user has to do is call `node` for the application to run?

**Comment by jviotti** [comment]
I think the problem is that the REPL code is implemented in C++, and not as a v8-based JavaScript program that we can replace. @RaisinTen might be able to confirm.

**Reply by aduh95** [reply]
REPL is a JS program: https://github.com/nodejs/node/blob/2178f1772f4f061e780780e99c709f94eb4b96cf/lib/internal/main/repl.js
C++ is involved to select what JS entry point to use (https://github.com/nodejs/node/blob/98fa488066b92a257d7d37f4ed24804850277594/src/node.cc#L505-L507), it could easily be replaced at compile time.

**Reply by jviotti** [reply]
Oh, interesting. I didn't know that. I have many questions then @RaisinTen:

- How are these files bundled in the Node.js executable?
- Does the REPL consist of a single file or multiple of them? If it's the latter, how are they bundled together?
- Is there any size/offset definition in the injected files that we can change and replace?

**Reply by RaisinTen** [reply]
> How are these files bundled in the Node.js executable?

It's done by js2c - https://github.com/nodejs/node/blob/472edc775d683aed2d9ed39ca7cf381f3e7e3ce2/tools/js2c.py. If you do a release build of node, you'll see `out/Release/obj/gen/node_javascript.cc` which contains a variable named `internal_main_repl_raw` that is assigned the contents of the `lib/internal/main/repl` file as a `static const uint8_t` array. The other occurrence of the variable in this file is the one which does `source_.emplace("internal/main/repl", UnionBytes{internal_main_repl_raw, 1842});`, to assign this array to the `"internal/main/repl"` key in `BuiltinSourceMap` - https://github.com/nodejs/node/blob/472edc775d683aed2d9ed39ca7cf381f3e7e3ce2/src/node_builtins.h#L125, so that it can be required internally using `require('internal/main/repl')`.

> Does the REPL consist of a single file or multiple of them? If it's the latter, how are they bundled together?

I would say multiple files because it requires some key components like `evalScript` (https://github.com/nodejs/node/blob/7f5679b901b28c9dd330a4f7391da626602452f2/lib/internal/process/execution.js#L53-L104) which is required from `internal/process/execution` in here - https://github.com/nodejs/node/blob/7f5679b901b28c9dd330a4f7391da626602452f2/lib/internal/main/repl.js#L12-L14.

> Is there any size/offset definition in the injected files that we can change and replace?

Injecting application code here would be the same as making changes to the source strings in the binary because there's no documented size/offset we can change to do this in the current state AFAIK. Another point to note here is that if we inject the application code here directly, it would be able to access the internal files like `require('internal/main/repl')`, so that would require us to run a bootstrap script which modifies `fs` to read from a vfs and then do something like `require('internal/modules/cjs/loader').Module.runMain(process.execPath);` (https://github.com/nodejs/node/blob/7f5679b901b28c9dd330a4f7391da626602452f2/lib/internal/main/run_main_module.js#L23) to set up the `require()` function that gets exposed to userland and finally run the application.

**Comment by guest271314** [comment]
@aduh95 Then this should be as simple as https://github.com/beaufortfrancois/download-chromium where `node` nightly source is used as the base source, user navigates to the page, uploads their JavaScript file(s), repl.js is replaced with the file(s), compiled on Google App Engine or comparable methods/means, code signed programatically https://github.com/nodejs/single-executable/discussions/14#discussioncomment-3400276

> Because code signatures operate on the binary format, any data section outside of the boundaries of the format is either not protected by the signature at all, or the OS will simply not run it (in the case of macOS)

and the `node` executable is output for user to download.

**Reply by RaisinTen** [reply]
> Can't we just substitute application code for built-in repl code?

The node binaries are probably already going to have the logic to detect whether application code has been injected into it and execute either that or the repl code. That's what https://github.com/nodejs/node/pull/42334 does already.

>  compiled on Google App Engine

That's definitely a possibility. If `node --compile` or something like that gets implemented anyone could create such a thing but I'm not sure why that would be useful.

**Reply by guest271314** [reply]
What I mean is just substitute/replace the repl cofe (JavaScript text) in the existing executable with arbitrary code supplied by user (JavaScript text). That way the user just call `node` and their replaced JavaScript cide is executed by `node`, just as calling `node` on command line starts repl.

**Comment by cspotcode** [comment]
What happens when they do `node foo.txt` does the repl still run?  Or is `foo.txt` executed as JS?

**Reply by jviotti** [reply]
Not sure about the proposed approach in this thread, but the overall SEA idea would be to pass `foo.txt` as the first argument to the program embedded into the executable.

**Comment by jesec** [comment]
There are many concerns with the approach. The ones I can think of right now:

- Command arguments handling. (we want `node app.js -- --arg` -> `./sea --arg`)
- Baking in and handling Node options. Some options must be loaded in the C++ world, before we have V8.
- REPL environment is not the same as normal. You may have noticed that you can use many built-ins in REPL without explicitly importing them.
- It is challenging to properly manipulate existing data/section in the binary post-link. Different compilers/linkers could embed data in different ways, and a marker is not necessarily present. We don't have tools to do that automatically and  deterministically for all required platforms. 

For the first 3 points, some may argue that logics can be added to handle the case, but that brings the complexity to the parallel of what we are already working on.

**Reply by ljharb** [reply]
also, a SEA should never be affected by the NODE_OPTIONS env var, or any node flags that the SEA author did not explicitly intend to be allowed.

**Reply by jesec** [reply]
Indeed. We want to allow users (developers who pack the SEA) to bake Node options into the single executable. End users (users who run the SEA) shall not have control over Node options (via NODE_OPTIONS env var or anything else).

**Reply by guest271314** [reply]
If the JavaScript is in the `node` executable I'm relatively sure I can just use the browser to modify the file, something like 

```
fetch('node')
.then((response) ==> response.text())
.then((text) => text.replace(/regex/,'const sea = []'))
.then((text) => new TextEncoder().encode(text))
.then((uint8) => showSaveFilePicker())
.then((handle)) => { /* write file */})
```

Done.

> There are many concerns with the approach. The ones I can think of right now:
>
> - Command arguments handling. (we want `node app.js -- --arg` -> `./sea --arg`)

I thought the whole point was making a single executable application, not passing arguments to the executable whatsoever. 

https://github.com/nodejs/single-executable/discussions/16#discussioncomment-3401898

> I don't think it counts as a single executable if you have to invoke it with specific parameters. Otherwise any zero-dependency node program could qualify merely by using `-e`.

We might as well just call `node example.js` if we are going to pass parameters to the executable. We've done nothing different from the EU perspective, I'm speaking for myself. Doesn't really make sense to me to pass arguments to a single-executable - it only does 1 thing. You want a different executable, bake the parameter into the executable and be done with it.

> - Baking in and handling Node options. Some options must be loaded in the C++ world, before we have V8.

Well, we can do that too. The whole `node` options is just what we already have. Nothing novel about a single executable that behaves the same way as `node` exeuctable, or existing Node.js patterns. For me the idea is to use as little `node` as possible - the single executable. 

> - REPL environment is not the same as normal. You may have noticed that you can use many built-ins in - REPL without explicitly importing them.
> It is challenging to properly manipulate existing data/section in the binary post-link. Different compilers/linkers could embed data in different ways, and a marker is not necessarily present. We don't have tools to do that automatically and deterministically for all required platforms.

I'm saying take the existing executable, and abuse it to make it do what you want after the fact, by simple manipulation of text in the executable. 

If we are going to customize `node` as a base source single executable application, we might as well just create a form on Node.js download Website, user fills in what they want, upload their application, and Node.js WG/Team, whoever is in charge of that domain, spits out the `node` executable customized: custom `node` builds. Done.

What I am reading appears to be a circular argument for unlimited options and indeed passing options to a single executable - leading right back to why not just use `node` executable and pass file name?

**Reply by ovflowd** [reply]
> We might as well just call node example.js if we are going to pass parameters to the executable. We've done nothing different from the EU perspective, I'm speaking for myself. Doesn't really make sense to me to pass arguments to a single-executable - it only does 1 thing. You want a different executable, bake the parameter into the executable and be done with it.

Isn't this a little bit too restrictive?

**Reply by jviotti** [reply]
> Doesn't really make sense to me to pass arguments to a single-executable - it only does 1 thing

The idea is to support building standalone CLIs built using Node.js. Almost every useful CLI out there does take arguments/options.

**Reply by ovflowd** [reply]
> The idea is to support building standalone CLIs built using Node.js. Almost every useful CLI out there does take arguments/options.

Exactly!

**Comment by guest271314** [comment]
For clarity, objectively this is the same; equivalent; no different:

> - Command arguments handling. (we want `node app.js -- --arg` -> `./sea --arg`)

The application needs to handle dynamics, if any. Remember the application is already built. It just happens to use `node` JavaScript runtime engine; the algorithm could easily use Python, C, C++, even Bash.

**Reply by cspotcode** [reply]
The `--` might cause some confusion by implying that node treats `--` specially when it does not.

**Comment by guest271314** [comment]
@cspotcode I don't think there is confusion. My understanding of a single executable application is no arguments/parameters need to be passed and are not expected to receive any parameters to the runtime itself. Ideally, we just call `node` with the custom `node` build incorporating the self-contained, non-dependent (on anything outside of the runtime executable and application) - else, the moment you pass environmental variables to `node` outside of the SEA, or pass arguments to the application side - the questions must be asked and answered: How is this different from `node example.js`, and therefore why not just use `node example.js`?

**Reply by ljharb** [reply]
The point of an SEA is that *nobody should have to, or be easily able to, have any idea whatsoever that the executable is built with node*. Anything that's node-specific should no longer exist, in other words, by the time the executable is created.

There's no value in providing a way to bundle node + a JS bundle in a single executable if the end user has to know how node works - if they know that, they already *have* node.

**Reply by ovflowd** [reply]
> There's no value in providing a way to bundle node + a JS bundle in a single executable if the end user has to know how node works - if they know that, they already have node.

I'd also agree with this. I would want to have a SEA distributed with Node.js to be as normal as any other *nix-like or Windows-like binary. Something I just execute without thinking about the specifics of Node.js environments, parameters or options.

**Comment by guest271314** [comment]
@ljharb I agree to some extent. I already have `node`, yet am not interested in `npm`, and packages. I have the same algorithm working in C, C++, Python, Bash, PHP, Node.js and QuickJS. While performance of v8 is faster than QuickJS per the record, you can't tell the difference between the source code when capturing and streaming audio (which is hard because any and all lags, etc. are perceptible in clipping, crackling, etc.).

I think we can achieve the goal here rather simply by providing a simple HTML form on the downloads Web site that accepts one or more files, writes the files to where repl would run with `node`. That does not mean use the repl, just use the same means that repl does to run `node` at command line, substituting the application therefore. Build `node` executable for the single user. Done.

**Reply by ljharb** [reply]
I’m pretty sure “node without npm or packages” isn’t a use case this repo is solving, and although you’re welcome to try separately, i doubt you’ll find a large vocal group who’s interested in that.

**Reply by guest271314** [reply]
That is my interest. I don't speak for others. And frankly, am not a follower. I don't really care what other people do. I am an individual. If I was to wait around for some absentee specification or implementation to do something I wouldn't be capturing entire system audio, specific audio, and speech synthesis output and parsing SSML to Web Speech API, or a local speech sythesis engine on Chromium/Chrome. "Do what thou wilt shall be the whole of the Law"-Thelema

**Reply by ovflowd** [reply]
> That is my interest. I don't speak for others. And frankly, am not a follower. I don't really care what other people do. I am an individual.

That's quite a rash way of thinking when you're imposing your voice to a (possible working-group) that is trying to cover a solution that would fit best the masses, not just our personal desires. 😅

**Reply by guest271314** [reply]
Those are the facts. I arrived at this destination after discovering Node.js does not just distribute the `node` executable. 

Some folks might be impressed by `npm` and N packages, I am not. Reminiscent of Microsoft shipping IE with Windows. You must like it. You must love it. Best thing ever!

Deno and bun provide executable only downloads. 

I just need the JavaScript runtime. Node.js does not just offer `node` executable for download. I wrote this https://gist.github.com/guest271314/012441b6809f60ea542b690660ee8e0d to fix that for myself.

I overtly notified you what my stake is. I have zero interest in `npm`, `npx` and whatever else is in the binary download besides `node` executable. Just don't need it. `node` executable JavaScript runtime doesn't need it either. People's feelings shouldn't be hurt because I have no use for their pakage manager. What is this about anyway, the package manager or the executable? The executable. Perhaps that is why Deno does not ship a package manager and extraneous folders and files in their download?

Again, the same question must ne asked once you try to defend `npm` and packages re passing arguments to a single executale application: If you are doing all that, what novel is going on in single-executable team? Why not just use `node example.js` and use `npm` to fetch a slew of packages?

> a solution that would fit best the masses, not just our personal desires. 

You are correct. My voice. Otherwise, what is the point of having a team if you do not actually listen to the members' concerns? You think you are on high and know what is best all of the time - for everybody. You don't. I'm advising you directly you don;t know what is best for me - else Node.js would provide a executable-only download without all the stuff you folks think is the bees knees.

Don't be a myopic organization that cannot handle internal criticism, that marginalizes folks who question what others claim to be what the masses want. That is precisely why Godel's Incompleteness Theorums are mathematically proven: You cannot verify an axiom of your system from inside your own system.

> I'm not interested in all that
> No thank you, I'll pass
> Please skip me for the next person
> I do appreciate that
> So kind of you to offer
> But no really I'm cool
> I try to stay away from the things that everybody else do
> I got my own thing happening
> It's working out great for me
> I'm healthy, I'm happy
> I'm successful with good money
> I hope that works out for you
> Don't let my good heart fool you
> I'll spazz out on you
> 
> -As if, Prodigy of Mobb Deep

**Reply by ovflowd** [reply]
@guest271314 there's no need to be rude here. I was never defending `npm` here, just mentioning that I suppose the interests of this project are for the greater good, not necessarily just our own.

**Reply by guest271314** [reply]
I'm not being "rude". I am not in awe either. This is direct communication. Who made you custodian of the greater good? Here I am providing my stake and you are trying to defend `npm`. Try listening to what I said. Your package manangement system is not the best ever and you should provide a `node`-only executable, like the masses of JavaScript runtime's do. The only way that is going to happen, apparently, is via this route, where I can insert the Native Messaging host into the executable and be done. That is the point here, right, a single-executable, not a praise moment for all thinkgs you think are great about Node.js.

**Reply by aduh95** [reply]
@guest271314 please edit your comments to remove personal attacks. No one here is defending npm, you are the only one who brings it up. It's OK to express your opinion and state your interests, but respect must come first.

**Reply by guest271314** [reply]
@ovflowd Again, the hallmark of a functional organization, institution, or team is how you handle internal criticism. It is easy enough to just dismiss ideas and interests that are not in-line with orthodoxy. That is democracy - majority rule, minority just has to deal with whatever hierarchy says. Take a moment to listen to constructive feedback.

**Reply by guest271314** [reply]
@aduh95 There were no personal attacks. I provided constructive criticism and the reason I discovered this effort. I cannot revise the truth.

**Reply by ovflowd** [reply]
> Again, the hallmark of a functional organization, institution, or team is how you handle internal criticism. 

I believe you misunderstood me. I apologise if my communication wasn't clear enough. I never defended or contradicted your criticism; I only mentioned that (what I believe) the goals of this project are for what kind of SEA solutions we can deliver that attend to the needs of the community, which of course, include yours, but not ultimately delimited/limited by only yours.

_Sorry for the off-topic_

**Reply by aduh95** [reply]
@guest271314 I'm not asking you to revise the truth, I'm asking you to communicate in a way that does not make the discussion toxic. This is your last warning, if you refuse to self-moderate, I'm going to ask @nodejs/moderation to block you from the Node.js org.

**Reply by guest271314** [reply]
@aduh95 You are mistaking direct communication, without rancor, for disrepect. I learned long ago that "respect" is just an emotion that makes an individual reliant on someone else's emotions, which can change at any moment. This is what is called diplomacy. You must be capable of handling internal constructive criticism to avoid being a tyranny from on high. 

I just want the `node` executable with my code inserted - that is the mission of this team. I said I don't need `npm`, which is consistent with that mission. The mere idea that somebody doesn't need `npm` and has no interest in it is not disrepect.

**Reply by guest271314** [reply]
> I don't like your tone here

Excuse me? My tone?

How exactly do you measure "tone"? I havn't called anybody out of their names, I don't do that. 

I've posted my ideas for SEA here. At some point I said I don't need and have no interest in `npm`, and that is rubbing people the wrong way?

For the record, my "tone" is one of an individual who has fought the state in court. You cannot fight the state and be meek and humble. You must assert yourself. 

I appreciate what you folks are doing here. Give and take. Do you really want diversity, or are you incapable of really handling diversity, which means you have to listen to people who you do not agree with?

I am more than willing to just post code - my experiments and tests to solve this issue - for the masses and for unique use cases. There is more than 1 use case for SEA. I shared mines, and why that is.

I'll just fous on the code. Have a great day!
